File: Napakalakiruby/lib/player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "dice.rb"
require_relative "treasure.rb"
require_relative "monster.rb"
require_relative "treasure_kind.rb"
require_relative "combat_result.rb"


class Player
  
  attr_accessor :name,:level,:dead,:canISteal,:enemy,:visibleTreasures,:hiddenTreasures,:pendingBadConsequence
  @@MAXLEVEL=10
  
  def initialize(nm)
    @name=nm
    @level=1
    @dead=true
    @visibleTreasures=Array.new 
    @hiddenTreasures=Array.new
  end
   
  
  def isDead
      @dead
  end
   
  def combat(m)
       myLevel=getCombatLevel  #1.1.1
       monsterLevel=m.getCombatLevel  #1.1.2
       
    if(myLevel>monsterLevel)
      applyPrize(m)  #1.1.3
      
      if(@level>=@@MAXLEVEL)
        result=CombatResult::WINGAME
      else
        result=CombatResult::WIN
      end
      
    else
      applyBadConsequence(m)  #1.1.4
      
      result=CombatResult::LOSE
    end
    
    return result   #1.1.10
  end
  
  def getHiddenTreasures
    @hiddenTreasures
  end
  
  def getVisibleTreasures
    @visibleTreasures
  end
  
  
  def makeTreasureVisible(t)
       canI=canMakeTreasureVisible(t) #1.2.1
       if(canI)
         @visibleTreasures<<t #1.2.2
         @hiddenTreasures.delete(t) #1.2.3
       end
  end 
   
  def discardVisibleTreasures( t)
    @visibleTreasures.delete(t) #1.2.1
    if((@pendingBadConsequence!=nil)&&(!@pendingBadConsequence.isEmpty)) 
      @pendingBadConsequence.substractVisibleTreasure(t)  #1.2.2
    end
    dieIfNotTreasures   #1.2.3
  end
   
  def discardHiddenTreasures( t)
      @hiddenTreasures.delete(t) #1.2.1
    if((@pendingBadConsequence!=nil)&&(!@pendingBadConsequence.isEmpty)) 
      @pendingBadConsequence.substractHiddenTreasure(t)  #1.2.2
    end
    dieIfNotTreasures   #1.2.3
  end
   
  def validState
    res=false
    if(@pendingBadConsequence.isEmpty&&@hiddenTreasures.length<4)
      res=true
    end
    return res
  end
   
  def initTreasures
    dealer=CardDealer.instance #1
    dice=Dice.instance #2
    bringToLife #3
    treasure=dealer.nextTreasure #4
    @hiddenTreasures<<treasure #5
    number=dice.nextNumber #6
    
    if(number>1)
      treasure=dealer.nextTreasure #7
      @hiddenTreasures<<treasure #8
    end
    
    if(number==6)
      treasure=dealer.nextTreasure #9
      @hiddenTreasures<<treasure #10
    end
  end
   
  def stealTreasure
    treasure=nil
    canI=canISteal
    if(canI)
      canYou=@enemy.canYouGiveMeATreasure
      if(canYou)
        treasure=enemy.giveMeATreasure
        @hiddenTreasures<<treasure
        haveStolen
      end
    end
    return treasure
  end
   
   
  def setEnemy(e)
       @enemy=e
  end
   
 def canISteal
     @canISteal
  end
   

   
  def discardAllTreasures
    visiblet=Array.new
    hiddent=Array.new
    visible=Array.new(@visibleTreasures)
    hiddent=Array.new(@hiddenTreasures)
    
       visiblet.each do |vt|
         discardVisibleTreasures(vt) #1.2
       end
       
    hiddent.each do |ht|
      discardHiddenTreasures(ht) #1.4
    end
  end
  
  def
    to_s
    "Nombre: #{@name} Nivel : #{@level}"
  end
    
  private
  
  def bringToLife
       @dead=false
  end
   
def  getCombatLevel
     lv=@level
     visibleTreasures.each {|t| lv+=t.bonus}
     return lv
  end
   
def incrementLevels(i)
      @level+=i
  end
   
 def decrementLevels(i)
       @level-=i
       if @level<1
         @level=1
       end
  end
   
def setPendingBadConsequence(b)
      @pendingBadConsequence=b 
  end
   
 def applyPrize( m)
   nLevels=m.getLevelsGained
   incrementLevels(nLevels)
   nTreasures=m.getTreasuresGained
   if(nTreasures>0)
     dealer=CardDealer.instance
     
     nTreasures.times do
       treasure=dealer.nextTreasure
       @hiddenTreasures<<treasure
     end
   end
  end
  
 def applyBadConsequence(m)
   badConsequence=m.getBadConsequence #1
   nLevels=badConsequence.getLevels #2
   decrementLevels(nLevels) #3
   pendingBad=badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)#4
   setPendingBadConsequence(pendingBad) #5
 end
   
def canMakeTreasureVisible(t)

            can=false
            
            nArmor=0
            nShoes=0
            nOnehand=0
            nBothhands=0
            nHelmet=0
            
            @visibleTreasures.each do |tr|  
                if(tr.type == TreasureKind::ARMOR)
                    nArmor=nArmor+1
                elsif(tr.type == TreasureKind::HELMET)
                    nHelmet=nHelmet+1
                elsif(tr.type == TreasureKind::ONEHAND)
                    nOnehand=nOnehand+1
                elsif(tr.type == TreasureKind::BOTHHAND)
                    nBothhands=nBothhands+1
                elsif(tr.type == TreasureKind::SHOES)
                    nShoes=nShoes+1
                end
            end

            if(t.type == TreasureKind::ARMOR && nArmor == 0)
                can = true
            elsif(t.type == TreasureKind::SHOES && nShoes == 0) 
                can = true
            elsif(t.type == TreasureKind::HELMET && nHelmet == 0)
                can = true
            elsif(t.type == TreasureKind::BOTHHAND && nBothhands == 0 && nOnehand== 0)
                can= true
            elsif(t.type == TreasureKind::ONEHAND && nOnehand < 2 &&nBothhands == 0)
                can = true
            end

            return can
  end
    
 def howManyVisibleTreasures(tKind)
     i=0
     @visibleTreasures.each {|t| if t.type ==tkind then i=il+1   end}
     return i
  end
   
 def dieIfNotTreasures
    if(@visibleTreasures.length==0&&@hiddenTreasures.length==0)
      dead=true
    end
  end
  
 def giveMeATreasure
   indice=rand(@hiddenTreasures.length)
   return @hiddenTreasures[indice]
  end
 
 def canYouGiveMeATreasure
       res=false
       if(@hiddenTreasures.length>0 && @visibleTreasures.length>0)
         res=true
       end
       res
  end
   
 def haveStolen
       @canISteal=false
  end
end

File: Napakalaki/src/napakalaki/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;

/**
 *
 * @author chen__000
 */
public class Player {
    
   final int MAXLEVEL=10;
   private String name;
   private int level;
   private boolean dead=true;
   private boolean canISteal=true;
   protected ArrayList<Treasure> hiddenTreasures;
   protected ArrayList<Treasure> visibleTreasures;
   private BadConsequence pendingBadConsequence;
   protected Player enemy;
   
   public Player(String nm){
       name=nm;
       level=1;
       hiddenTreasures=new ArrayList();
       visibleTreasures=new ArrayList();
       this.pendingBadConsequence= new NumericBadConsequence("",0,0,0);
   }
   
   public Player(Player p){
       this.name=p.name;
       this.level=p.level;
       this.hiddenTreasures=p.hiddenTreasures;
       this.visibleTreasures=p.visibleTreasures;
       this.pendingBadConsequence=p.pendingBadConsequence;
   }
   
   public String getName(){
       return name;
   }
   
   private void bringToLife(){
       dead=false;
   }
   
   protected int getCombatLevel(){
       int lv=level;
       for(Treasure t: visibleTreasures){
           lv=lv+t.getBonus();
       }
        return lv;
   }
   
   private void incrementLevels(int i){
       level=level+i;
   }
   
   private void decrementLevels(int i){
       level=level-i;
       if(level<1){
           level=1;
       }
   }
   
   private void setPendingBadConsequence(BadConsequence b){
       pendingBadConsequence=b;
   }
   
   private void applyPrize(Monster m){
       int nLevels=m.getLevelsGained(); //1
       this.incrementLevels(nLevels);  //2
       int nTreasures=m.getTreasuresGained(); // 3
       
       if(nTreasures>0){
           CardDealer dealer=CardDealer.getInstance();  //4
       
          for(int i=0;i<nTreasures;i++){
             Treasure treasure=dealer.nextTreasure();  //5
             this.hiddenTreasures.add(treasure);  //6
           }
       }
   }
   
   private void applyBadConsequence(Monster m){
       BadConsequence badConsequence=m.getBadConsequence();//1
       int nLevels = badConsequence.getLevels();
       this.decrementLevels(nLevels);//3
       BadConsequence pendingBad=badConsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);//4
       this.setPendingBadConsequence(pendingBad); //5
   }
   
   private boolean canMakeTreasureVisible(Treasure t){
       int nOnehand=0;
       int nBothhands=0;
       int nArmor=0;
       int nShoes=0;
       int nHelmet=0;
       boolean res=false;
       
       for (Treasure tre : visibleTreasures) {
           if(tre.getType() == TreasureKind.ONEHAND){
               nOnehand++;
           }else if(tre.getType() ==  TreasureKind.BOTHHANDS){
               nBothhands++;
           }else if(tre.getType() == TreasureKind.ARMOR){
               nArmor++;
           }else if(tre.getType() ==  TreasureKind.SHOES){
               nShoes++;
           }else{
               nHelmet++;
           }
       }
       switch(t.getType()){
            case ONEHAND:
                if(nOnehand < 2 && nBothhands == 0)
                    res = true;
                break;
            case BOTHHANDS:
                if(nOnehand == 0 && nBothhands == 0)
                    res = true;
                break;
            case HELMET:
                if(nHelmet == 0)
                    res = true;
                break;
            case SHOES:
                if(nShoes == 0)
                    res = true;
                break;
            case ARMOR:
                if(nArmor == 0)
                    res = true;
                break;
        }
       return res;
   }
    
   private int howManyVisibleTreasures(TreasureKind tKind){
       int i=0;
       for(Treasure t: visibleTreasures){
           if(t.getType()==tKind){
               i++;
           }
       }
       return i;
   }
   
   private void dieIfNotTreasures(){
       if(visibleTreasures.isEmpty()&&hiddenTreasures.isEmpty())
            dead=true;
   }
   
   public boolean isDead(){
       return dead;
   }
   
   public ArrayList<Treasure> getHiddenTreasures(){
       return this.hiddenTreasures;
   }
   
   public ArrayList<Treasure> getVisibleTreasures(){
       return this.visibleTreasures;
   }
   
   public CombatResult combat(Monster m){
       CombatResult result;
       boolean should;
       int myLevel=this.getCombatLevel();  //1.1.1
       int monsterLevel=this.getOponentLevel(m); //1.1.2
       
       if(myLevel>monsterLevel){
           this.applyPrize(m); //1.1.3
           
           if(this.level>=MAXLEVEL){
               result=CombatResult.WINGAME;
           }
           else{
               result=CombatResult.WIN;
           }
       }
       else{
           this.applyBadConsequence(m); //1.1.4
           
           should=this.shouldConvert();
           if(should==true){
               result=CombatResult.LOSEANDCONVERT;
           }
           else{
               result=CombatResult.LOSE;
           }
       }
       return result;
    }
   
   public void makeTreasureVisible(Treasure t){
       boolean canI=this.canMakeTreasureVisible(t); //1.2.1
       if(canI){
           this.visibleTreasures.add(t);  //1.2.2
           this.hiddenTreasures.remove(t);  //1.2.3
       }
   }
   
   public void discardVisibleTreasures(Treasure t){
       this.visibleTreasures.remove(t);  //1.2.1
       if((this.pendingBadConsequence!=null) && (!this.pendingBadConsequence.isEmpty())){ 
           this.pendingBadConsequence.substractVisibleTreasure(t);  //1.2.2
       }
      this.dieIfNotTreasures(); // 1.2.3
   }
   
   public void discardHiddenTreasures(Treasure t){
       this.hiddenTreasures.remove(t);  //1.2.1
       while((this.pendingBadConsequence!=null) && (!this.pendingBadConsequence.isEmpty())){ 
           this.pendingBadConsequence.substractHiddenTreasure(t);  //1.2.2
       }
      this.dieIfNotTreasures(); // 1.2.3
   }
   
   public boolean validState(){
       boolean res=false;
        if(pendingBadConsequence==null||pendingBadConsequence.isEmpty() && hiddenTreasures.size()<=4 ){
            res=true;
        } 
        return res;
   }
   
   public void initTreasures(){
       Treasure treasure;
       int number;
       CardDealer dealer=CardDealer.getInstance();  //1
       Dice dice=Dice.getInstance();//2
       this.bringToLife();//3
       treasure=dealer.nextTreasure();//4
       this.hiddenTreasures.add(treasure);//5
       number=dice.nextNumber();//6
       
       if(number>1){
           treasure=dealer.nextTreasure();//7
           this.hiddenTreasures.add(treasure);//8
       }
       if(number==6){
           treasure=dealer.nextTreasure();//9
           this.hiddenTreasures.add(treasure);//10
       }
   }
   
   public int getLevels(){
       return level;
   }
   
   public Treasure stealTreasure() {
       Treasure treasure=null;
       boolean canI=this.canISteal();  //1.1
       if(canI){
           boolean canYou=this.enemy.canYouGiveMeATreasure();  //1.2
           if(canYou){
               treasure=this.enemy.giveMeATreasure();  //1.3
               this.hiddenTreasures.add(treasure);  //1.4
               this.haveStolen();  //1.6
           }
       }
       return treasure;
   }
   
   
   public void setEnemy(Player e){
       enemy=e;
   }
   
   protected Treasure giveMeATreasure(){
        int numero= (int) (Math.random()*hiddenTreasures.size());
        return hiddenTreasures.get(numero);
   }
   
   public boolean canISteal(){
       return this.canISteal;
   }
   
   protected boolean canYouGiveMeATreasure(){
       boolean res=false;
       if(this.hiddenTreasures.size()>0){
           res=true;
       }
       return res;
   }
   
   private void haveStolen() {
        this.canISteal=false;
   }
   
   public void discardAllTreasures(){
       Treasure treasure;
       ArrayList<Treasure> visiblet=new ArrayList(this.visibleTreasures);
       ArrayList<Treasure> hiddent=new ArrayList(this.hiddenTreasures);
       
       for(Treasure vt : visiblet){
          this.discardVisibleTreasures(vt); //1.2
       }
       for(Treasure ht : hiddent){
           this.discardHiddenTreasures(ht);  //1.4
       }
   }
   
   
     protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        Dice dice=Dice.getInstance();
        boolean should=false;
        
        if(dice.nextNumber()==1){
            should=true;
        }
        
        return should;
    }
    
   
    public String toString() {
        return "Nombre: " + this.name + " Nivel: " + this.level;
    }
   
}
